### PR TITLE
DOC: Document/ Deprecate functions exposed in "numpy" namespace

### PR DESCRIPTION
**Pull request by kritisingh1**
Related to #12385 
The checklist for future work on this PR is as follows:
Deprecate - 
- [x] 'mafromtxt`
- [x] 'ndfromtxt'

@rgommers Please review and let me know the required changes. Thanks a lot!

### CODE REVIEW

**Comment by mattip** [comment]
Please remove from your checklist items that already appear in #13214.

**Comment by mattip** [comment]
xref #9312 which began documenting `show_config` but still needs work.

**Comment by kritisingh1** [comment]
I have added the warning messages for the function `ndfromtxt` and `mafromtxt` and also edited the docstrings. I am not sure where equivalent functions for `int_asbuffer` exists and would be grateful for pointers on it. @mattip

**Comment by seberg** [comment]
@kritisingh1 `int_asbuffer` is defined in `multiarray` on the C side. Sometimes a nice way to find things can be to use `git grep "int_asbuffer"`, etc. The definition is in `numpy/core/src/multiarray/multiarraymodule.c`.

**Comment by kritisingh1** [comment]
@seberg yeah I was using `git grep "int_asbuffer" but the function name was `asbuffer` so it didn't turn up in the results. I will take care of this from next time.

**Comment by mattip** [comment]
the interesting part will be determining what its docstring **should** be so we can put a deprecation notice in it. The docstring should be added to `numpy/core/_add_newdocs.py`, like 
```
add_newdoc('numpy.core.multiarray', 'int_asbuffer', '''
<your text here>
.. deprecated:: 1.17
     Deprecated, use <your text here> instead
 ''')
```
Play with the function a bit, read the C code, and ask here or on the mailing list if you have a hard time coming up with some text

File: numpy/lib/npyio.py [hunk 1]
```diff
@@ -2225,6 +2225,8 @@ def ndfromtxt(fname, **kwargs):
     """
     Load ASCII data stored in a file and return it as a single array.
 
+    .. deprecated:: np.ndfromtxt is deprecated
+
```
Comment (mattip) [inline numpy/lib/npyio.py]: needs a version number, see the other deprecation notices

File: numpy/lib/npyio.py [hunk 2]
```diff
@@ -2235,13 +2237,18 @@ def ndfromtxt(fname, **kwargs):
 
     """
     kwargs['usemask'] = False
+    warnings.warn(
+        "np.ndfromtxt is deprecated",
+        DeprecationWarning, stacklevel=2)
```
Comment (mattip) [inline numpy/lib/npyio.py]: Needs a comment above the warning with the date, version, gh-issue format

**Comment by mattip** [comment]
Please update the checklist at the top of the PR with the additional items, and mark off the ones that are modified in the PR

File: numpy/lib/npyio.py (hunk 1 above)
Comment (kritisingh1) [reply numpy/lib/npyio.py]: I was wondering if it should be 1.17?

**Comment by kritisingh1** [comment]
@mattip I'm figuring out what text to add to the `numpy/core/_add_newdocs.py` . Meanwhile, I would be grateful if you could help me in fixing the failing the tests. By the logs, the tests which are calling `safe_eval`, `ndfromtxt` and `mafromtxt` and are encountering `warnings.warn(..)` in the respective functions are failing which should not happen because they are just warnings. I'm not sure why this is happening. Thanks!

**Comment by mattip** [comment]
Look at the `with warnings.*` constructs in other tests, they are used to filter known warnings. Testing turns unexpected warnings into errors.

**Comment by mattip** [comment]
You should be running at least the basic tests locally, so you can avoid too much CI churn `python runtests.py`

**Comment by kritisingh1** [comment]
@mattip the tests are failing despite adding the `with warnings.*` constructor for every mention of the deprecated functions. Is there anything I am missing? 
Also, I looked at other functions which were deprecated in `npyio.py` and could see that the alternative to those functions were used in the tests. For example. `pickle.loads` in place of `loads`. So, as @rgommers suggested, I can instead modify the tests which use `ndfromtxt` and `mafromtxt` to use `genfromtxt` with the required flags as parameters. Please let me know what should be the better way to go ahead. Thanks!

**Comment by mattip** [comment]
You should add a test to `numpy/core/tests/test_deprecated.py` (or `numpy/ma/tests/test_deprecations.py` or `numpy/linalg/tests/test_deprecations.py`), for each deprecation. Then replace the deprecated function with its alternative inside numpy itself, so any function that used to call the deprecated function now should not be emitting warnings.
 
> tests are failing despite adding the with warnings.* construct

It is hard to tell without seeing your changes. Note `test_deprecations` uses `self.assert_deprecated`

File: numpy/lib/utils.py [hunk 3]
```diff
@@ -1136,7 +1139,6 @@ def safe_eval(source):
     """
     # Local import to speed up numpy's import time.
     import ast
-
```
Comment (seberg) [inline numpy/lib/utils.py]: I think you can put this deprecation back, but have to replace all use-cases with the `ast.literal_eval(source)` call that this does. Maybe the DeprecationWarning should mention to use `ast.literal_eval` instead.

Comment (kritisingh1) [reply numpy/lib/utils.py]: @seberg , can I try deprecating `self_eval` in a different pull request altogether, if that is alright?

File: numpy/lib/npyio.py (hunk 2 above)
Comment (eric-wieser) [reply numpy/lib/npyio.py]: Should also recommend a replacement, perhaps `` `ndfromtxt` is a deprecated alias of `genfromtxt`, prefer the latter ``

File: numpy/lib/npyio.py [hunk 4]
```diff
@@ -2225,6 +2225,9 @@ def ndfromtxt(fname, **kwargs):
     """
     Load ASCII data stored in a file and return it as a single array.
 
+    .. deprecated:: 1.17
+        np.ndfromtxt is deprecated
```
Comment (eric-wieser) [inline numpy/lib/npyio.py]: Again, mention the replacement here, possibly with a few more words than in the `DeprecationWarning`.
```rst
`ndfromtxt` is a deprecated alias of `genfromtxt` which overwrites the ``usemask`` argument with `False` even when explicitly called as ``ndfromtxt(..., usemask=True)``. Use `genfromtxt` instead.
```

File: numpy/lib/utils.py (hunk 3 above)
Comment (seberg) [reply numpy/lib/utils.py]: Yes of course, but should probably remove the documentation change as well for now then.

Comment (kritisingh1) [reply numpy/lib/utils.py]: done

**Comment by kritisingh1** [comment]
@mattip I will open separate PRs for addressing the issue #12385 as and when I complete the work involved for documenting/deprecating different functions. Meanwhile, please let me know if there is anything else that needs to be done for deprecating the functions in the above checklist.

**Comment by mattip** [comment]
Thanks @kritisingh1

**Pull request merged**